### PR TITLE
DetectNet layers and example

### PR DESCRIPTION
**Pull request by lukeyeager**

### CODE REVIEW

File: examples/kitti/detectnet_network.prototxt [hunk 1]
```diff
@@ -0,0 +1,2536 @@
...
+    scale_min: 0.8
+    scale_max: 1.2
+    hue_rotation_prob: 0.8
+    hue_rotation: 30
+    desaturation_prob: 0.8
+    desaturation_max: 0.8
+  }
+  transform_param: {
+    mean_value: 127
+  }
```
Comment (gheinrich) [inline examples/kitti/detectnet_network.prototxt:100]: don't you need to shift the input by 127 in the deploy network to reproduce this transform param? In my network I had this:

```
layer {
  name: "shift"
  type: "Power"
  bottom: "data"
  top: "transformed_data"
  include {
    stage: "deploy"
  }
  power_param {
    shift: -127.0
  }
}
```

I had also changed the top of the transformation layer to be `transformed_data` as I was using a version of DIGITS that required the data layer to have a top called `data` but maybe you can do the shift in place now.

Comment (gheinrich) [inline examples/kitti/detectnet_network.prototxt:100]: Or maybe we should remove this `mean_value` parameter? I measured the mean pixel on a popular object detection dataset at `93`. Not quite `127`... with DIGITS wouldn't it be better to do proper mean subtraction (based on the dataset mean not a constant).

**Comment by drnikolaev** [comment]
Could you check Travis build errors like
`src/caffe/layers/detectnet_transform_layer.cpp:93:56: error: ‘UINT8_MAX’ was not declared in this scope`

File: src/caffe/layers/l1_loss_layer.cpp [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+                                            const vector<Blob<Dtype>*>& top) {
+  int count = bottom[0]->count();
+  caffe_sub(
+      count,
+      bottom[0]->cpu_data(),
+      bottom[1]->cpu_data(),
+      diff_.mutable_cpu_data());
+  caffe_cpu_sign(count, diff_.cpu_data(), sign_.mutable_cpu_data());
+  Dtype abs_sum = caffe_cpu_asum(count, diff_.cpu_data());
+  Dtype loss = abs_sum / bottom[0]->num();
```
Comment (gheinrich) [inline src/caffe/layers/l1_loss_layer.cpp:33]: I think you should divide by `count` here to normalize the loss. By the way `num()` is deprecated - [see](http://caffe.berkeleyvision.org/doxygen/classcaffe_1_1Blob.html#a56c2b25db397d9e82bbd7c43597ae427)

File: examples/kitti/detectnet_network.prototxt (hunk 1 above)
Comment (lukeyeager) [inline examples/kitti/detectnet_network.prototxt:100]: > I was using a version of DIGITS that required the data layer to have a top called data but maybe you can do the shift in place now.

Yeah, that should work now that https://github.com/NVIDIA/DIGITS/pull/763 has been merged.

Depending on the implementation, maybe we can move the `transform_param` into the `Data` layer? Then for this Caffe example it would be clear that people should be subtracting the same value for deployment pipeline, and in DIGITS the mean value can be calculated automatically for each dataset?

File: src/caffe/layers/l1_loss_layer.cpp (hunk 2 above)
Comment (lukeyeager) [inline src/caffe/layers/l1_loss_layer.cpp:33]: /cc @drendleman

**Comment by lukeyeager** [comment]
> Could you check Travis build errors like
> `src/caffe/layers/detectnet_transform_layer.cpp:93:56: error: ‘UINT8_MAX’ was not declared in this scope`

I'm assuming this is another Ubuntu 12.04 issue? It builds fine on 14.04 and 16.04. Maybe I should do something like https://github.com/NVIDIA/DIGITS/pull/766 for NVcaffe?

File: examples/kitti/detectnet_network.prototxt (hunk 1 above)
Comment (lukeyeager) [inline examples/kitti/detectnet_network.prototxt:100]: @drendleman says some of the transformations work best / most simply when they get to access the data before doing mean subtraction. Since we can work around it as-is, I'm fine leaving it in this layer.

Comment (gheinrich) [inline examples/kitti/detectnet_network.prototxt:100]: If we leave the shift in this layer then we need to add a `Power` layer to shift data by the same amount in the deploy network, right? And we need to tell DIGITS users to disable mean subtraction?

**Comment by lukeyeager** [comment]
Rebased to include TravisCI changes in https://github.com/NVIDIA/caffe/pull/146. Now the OpenCV version is new enough to build these layers.

However, now we can see that the C++ layers fail to build when OpenCV is disabled with the `USE_OPENCV` flag:
https://travis-ci.org/lukeyeager/caffe/builds/132970345

**Comment by drendleman** [comment]
What's a good way to fix this? The DetectNet header exports OpenCV types

File: examples/kitti/detectnet_network.prototxt (hunk 1 above)
Comment (lukeyeager) [inline examples/kitti/detectnet_network.prototxt:100]: Good call Greg. I'm updating it to look like this:

detectnet-power-layer

**Comment by lukeyeager** [comment]
All tests are passing. I'll merge this and we can address any updates to the model or solver in subsequent PRs.

**Comment by willyd** [comment]
This DetectNet and the DIGITS support for it seems very interesting. Is there a scientific paper describing the method you use for detection? Is this similar to (Faster, Fast) R-CNN?
